ergm.py

```diff
@@ -136,72 +142,166 @@ def approximate_auto_correlation_function(self, networks_sample: np.ndarray) ->
+                batch
+                    based on difference of means of sample batches from the total mean, as in Geyer's handbook of
+                    MCMC (there it is stated for the univariate case, but the generalization is straight forward).
+                multivariate_initial_sequence
+                    Following Dai and Jones 2017 - the first estimator in section 3.1 (denoted mIS).
         TODO: implement a mechanism that allows to pass arguments that are customized for each method
 
         Returns
         -------
         The covariance matrix estimation (num_features X num_features).
         """
-        if method == 'batch':
+        if method == 'naive':
+            num_features = features_of_net_samples.shape[0]
+            sample_size = features_of_net_samples.shape[1]
+            mean_features = np.mean(features_of_net_samples, axis=1)
+            # An outer product of the means (E[gi]E[gj])
+            cross_prod_mean_features = (mean_features.reshape(num_features, 1) @
+                                        mean_features.T.reshape(1, num_features))
+            # A mean of the outer products of the sample (E[gi*gj])
+            mean_features_cross_prod = np.mean(
+                features_of_net_samples.T.reshape(sample_size, num_features, 1) @
+                features_of_net_samples.T.reshape(sample_size, 1, num_features), axis=0)
+
+            return mean_features_cross_prod - cross_prod_mean_features
+
+        elif method == 'batch':
             num_features = features_of_net_samples.shape[0]
             sample_size = features_of_net_samples.shape[1]
             # Verify that the sample is nicely divided into non-overlapping batches.
             while sample_size % num_batches != 0:
                 num_batches += 1
             sample_mean = features_of_net_samples.mean(axis=1)
             batch_size = sample_size // num_batches
-            
+
             # Divide the sample into batches, and calculate the mean of each one of them
             batches_means = np.zeros((num_features, num_batches))
             for i in range(num_batches):
-                batches_means[:, i] = features_of_net_samples[:, i*batch_size:(i+1)*batch_size].mean(axis=1)
+                batches_means[:, i] = features_of_net_samples[:, i * batch_size:(i + 1) * batch_size].mean(axis=1)
 
-                
             diff_of_global_mean = batches_means - sample_mean.reshape(num_features, 1)
-            
+
             # Average the outer products of the differences between batch means and the global mean
             batches_cov_mat_est = np.mean(
                 diff_of_global_mean.T.reshape(num_batches, num_features, 1) @
                 diff_of_global_mean.T.reshape(num_batches, 1, num_features), axis=0)
             # Multiply by the batch size to compensate for the aggregation into batches.
             return batch_size * batches_cov_mat_est
+
+        elif method == "multivariate_initial_sequence":
```

Comment:
I guess we're not doing the "adjusted multivariate initial sequence" version because it's too slow?

But anyway I'm not 100% sure I understand how it helps. They say that adding Gammas with a negative eigenvalue "will squeeze the corresponding confidence region in undesirable directions". So by forcing a matrix with all positive eigenvalues, don't we immediately get the "index" we want to cut?